AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -159,22 +160,49 @@ - (void)_layoutNodes:(NSArray *)nodes ofKind:(NSString *)kind atIndexPaths:(NSAr
   if (!nodes.count) {
     return;
   }
-  
+
+  NSUInteger nodeCount = nodes.count;
+  NSMutableArray<ASCellNode *> *allocatedNodes = [NSMutableArray arrayWithCapacity:nodeCount];
   dispatch_group_t layoutGroup = dispatch_group_create();
-  ASSizeRange *nodeBoundSizes = (ASSizeRange *)malloc(sizeof(ASSizeRange) * nodes.count);
+  ASSizeRange *nodeBoundSizes = (ASSizeRange *)malloc(sizeof(ASSizeRange) * nodeCount);
   for (NSUInteger j = 0; j < nodes.count && j < indexPaths.count; j += kASDataControllerSizingCountPerProcessor) {
     NSInteger batchCount = MIN(kASDataControllerSizingCountPerProcessor, indexPaths.count - j);
-    
-    for (NSUInteger k = j; k < j + batchCount; k++) {
-      ASCellNode *node = nodes[k];
-      if (!node.isNodeLoaded) {
-        nodeBoundSizes[k] = [self constrainedSizeForNodeOfKind:kind atIndexPath:indexPaths[k]];
+
+
+    if (ASDisplayNodeThreadIsMain()) {
+      dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        for (NSUInteger k = j; k < j + batchCount; k++) {
```

Comment:
Should this code be defined in the block, provided directly to dispatch a sink here, while we just call it in line in the else condition below? I guess the dispatch semaphore signaling is the main difference, maybe it is not worth trying to figure out how to de-duplicate that but we could use a Boolean variable to only call it in the case that is needed, although it would be a bit weird for the sema variable to be NULL in that case it would probably be better than the duplication.